### PR TITLE
`Iris`: Fixed an issue where the send button would be disabled when a step execution error appears

### PR DESCRIPTION
**Pull request by RY997**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [ ] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [ ] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [ ] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [ ] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context

When the execution step fails and a detailed error message is displayed at the bottom of the chat window, the user cannot use the send button and clicking the retry button does not remove the error message. The error message disappears and the send button is enabled only when the user reloads the page.

### Description

The error information of step execution is not stored in the state storage, but the detailed error information is bound to the specific step and displayed below the step.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise with Code Editor Iris enabled, you can use [exercise 10](https://artemis-test9.artemis.cit.tum.de/course-management/11/programming-exercises/54/code-editor/179)

1. Log in to Artemis
2. Navigate to "edit in editor" of an exercise
3. Chat with Iris, when a step failed, the error should be displayed below the step and the send button should not be disabled.

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

### Screenshots

Before
Screenshot 2023-11-30 at 14 32 57

After
Screenshot 2023-11-30 at 14 34 48

## Summary by CodeRabbit

- **New Features**
	- Enhanced chatbot widget with improved error handling and user feedback on step failures.
- **Bug Fixes**
	- Adjusted timeout settings to optimize chatbot responsiveness.
- **Tests**
	- Updated chatbot widget tests for better reliability and maintainability.

### CODE REVIEW

**Comment by Ceiridge** [comment]
The send button is still disabled when an error appears, in this case that there is no API token:
image

So: Not working as expected on ts9

**Comment by RY997** [comment]
> The send button is still disabled when an error appears, in this case that there is no API token: image
> 
> So: Not working as expected on ts9

This is not included in this PR and it's the expected behavior

**Review by laurenzfb — APPROVED** [review]
tested in testing session on ts9. Managed to get an error and it was displayed as described. Code also LGTM

**Review by reschandreas — APPROVED** [review]
tested in testing session on TS9, works as expected.

**Review by mateusmm01 — APPROVED** [review]
Tested during testing session on ts9. Works as expected!


Screenshot 2023-12-21 at 15 40 04
Screenshot 2023-12-21 at 15 40 14

File: src/main/webapp/app/iris/exercise-chatbot/widget/chatbot-widget.component.ts [hunk 1]
```diff
@@ -771,9 +772,9 @@ export class IrisChatbotWidgetComponent implements OnInit, OnDestroy, AfterViewI
         plan.executing = false;
         step.executionStage = ExecutionStage.FAILED;
         if (!errorTranslationKey) {
-            this.stateStore.dispatch(new ConversationErrorOccurredAction(IrisErrorMessageKey.TECHNICAL_ERROR_RESPONSE));
+            this.stepError = this.translateService.instant(IrisErrorMessageKey.TECHNICAL_ERROR_RESPONSE);
         } else {
-            this.stateStore.dispatch(new ConversationErrorOccurredAction(errorTranslationKey, translationParams));
+            this.stepError = this.translateService.instant(errorTranslationKey, translationParams);
```
[the remark this answers is not in the document]
Comment (RY997) [reply src/main/webapp/app/iris/exercise-chatbot/widget/chatbot-widget.component.ts:777]: step error messages are not supposed to be dispatched to the state store, that's the main change for this PR

File: src/main/webapp/app/iris/exercise-chatbot/widget/chatbot-widget.component.ts [hunk 2]
```diff
@@ -365,7 +366,7 @@ export class IrisChatbotWidgetComponent implements OnInit, OnDestroy, AfterViewI
                 // will be cleared by the store automatically
                 this.stateStore.dispatch(new ConversationErrorOccurredAction(IrisErrorMessageKey.IRIS_SERVER_RESPONSE_TIMEOUT));
                 this.scrollToBottom('smooth');
-            }, 20000);
+            }, 60000);
```
[the remark this answers is not in the document]
Comment (RY997) [reply src/main/webapp/app/iris/exercise-chatbot/widget/chatbot-widget.component.ts:369]: It will be cleared by the state store

**Review by MaximilianJG — APPROVED** [review]
For me it always executes successfully. Is this intended? 

Screenshot 2024-01-09 at 21 56 57

**Comment by RY997** [comment]
> For me it always executes successfully. Is this intended?
> 
> Screenshot 2024-01-09 at 21 56 57

Yes, this is the expected behavior of the feature and this PR is to improve the way errors are handled when execution fails. If it's successful, then it can not be tested.

**Pull request merged**